File: 322-coin-change/322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
       int[] change = new int[amount+1];
    Arrays.fill(change,-1);
    change[0]=0;
    
    for(int i=1;i<=amount;i++){
        boolean flag = false;
        int min = Integer.MAX_VALUE;
        for(int coin:coins){
            if(i>=coin && change[i-coin] != -1){
                flag=true;
                min = Math.min(min,change[i-coin]);
            }
        }
        if(flag){
            change[i]= 1+min;
        }
    }
    return change[amount];  
    }
}

File: 322-coin-change/322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
    vector<int> need(amount+1, amount+1);
    need[0] = 0;
    for (int c : coins){
         //cout<<" c = "<<c<<endl; 
        for (int a=c; a<=amount; a++)
        {  need[a] = min(need[a], need[a-c] + 1);
         //cout<<" need[a] = "<<need[a]<<endl; 
        }
    }
    return need.back() > amount ? -1 : need.back();
    }
};